src/test/java/bio/terra/service/snapshotbuilder/SnapshotBuilderServiceTest.java

```diff
@@ -590,4 +593,99 @@ void notifySnapshotReady() {
         .snapshotReady(
             eq(id), anyString(), eq(snapshot.getName()), eq("No snapshot specification found"));
   }
+
+  @Nested
+  class GroupMemberApis {
+
+    SnapshotAccessRequestModel requestModel =
+        SnapshotBuilderTestData.createAccessRequestModelSnapshotCreated();
+    UUID requestId = new SnapshotAccessRequestResponse().getId();
+
+    @BeforeEach
+    void beforeEach() {
+      when(snapshotRequestDao.getById(requestId)).thenReturn(requestModel);
+    }
+
+    @Nested
+    class BlankSamGroup {
+      SnapshotAccessRequestModel requestModelBlankGroup =
+          SnapshotBuilderTestData.createAccessRequestModelApproved();
+      UUID requestIdBlankGroup = new SnapshotAccessRequestResponse().getId();
+
+      @BeforeEach
+      void beforeEach() {
+        when(snapshotRequestDao.getById(requestIdBlankGroup)).thenReturn(requestModelBlankGroup);
+      }
+
+      @Test
+      void testGetGroupMembersNotCreated() {
+        assertThrows(
+            IllegalArgumentException.class,
+            () -> snapshotBuilderService.getGroupMembers(requestIdBlankGroup));
+      }
+
+      @Test
+      void testAddGroupMemberNotCreated() {
+        String memberEmail = "user@gmail.com";
+        assertThrows(
+            IllegalArgumentException.class,
+            () -> snapshotBuilderService.addGroupMember(requestIdBlankGroup, memberEmail));
+      }
+
+      @Test
+      void testDeleteGroupMemberNotCreated() {
+        String memberEmail = "user@gmail.com";
+        assertThrows(
+            IllegalArgumentException.class,
+            () -> snapshotBuilderService.deleteGroupMember(requestIdBlankGroup, memberEmail));
+      }
+    }
+
+    @Test
+    void testGetRequestGroupMembers() {
+      SnapshotAccessRequestMembersResponse expected =
+          new SnapshotAccessRequestMembersResponse().members(List.of());
+      when(iamService.getGroupPolicyEmails(requestModel.samGroupName(), IamRole.MEMBER.toString()))
+          .thenReturn(expected.getMembers());
+      assertThat(snapshotBuilderService.getGroupMembers(requestId), is(expected));
+    }
+
+    @Test
+    void testAddRequestGroupMember() {
+      String memberEmail = "user@gmail.com";
+      SnapshotAccessRequestMembersResponse expected =
+          new SnapshotAccessRequestMembersResponse().members(List.of(memberEmail));
+      when(iamService.addEmailToGroup(
+              requestModel.samGroupName(), IamRole.MEMBER.toString(), memberEmail))
+          .thenReturn(expected.getMembers());
+      assertThat(snapshotBuilderService.addGroupMember(requestId, memberEmail), is(expected));
+    }
+
+    @Test
+    void testAddRequestGroupMemberInvalidEmail() {
+      String badEmail = "badEmail";
```

Comment:
It's a bit annoying that we have to test each validation case separately for each API. If the validation was refactored into a separate method, you could test that method once for each validation case instead. Something like
```
class SnapshotBuilderService {
...
  static void validateGroupParams(SnapshotAccessRequestModel model, String memberEmail) {
    if (memberEmail != null) {
      ValidationUtils.requireValidEmail(memberEmail, "Invalid member email");
    }
    ValidationUtils.requireNotBlank(
        model.samGroupName(),
        "Snapshot must be created from this request in order to manage group membership.");
  }
```

Reply:
I followed this suggestion and subsequently removed the BlankSamGroup class from the tests and made a new class for validating parameters that tests multiple cases of this method (blank group, bad email, etc.)